launcher/src/main/java/org/apache/spark/launcher/SparkLauncher.java

```diff
@@ -0,0 +1,371 @@
+    return appThread;
+  }
+
+  /**
+   * Launches a sub-process that will start the configured Spark application.
+   *
+   * @return A process handle for the Spark app.
+   */
+  public Process launch() throws IOException {
+    List<String> cmd = buildLauncherCommand();
+    return Runtime.getRuntime().exec(cmd.toArray(new String[cmd.size()]));
+  }
+
+  SparkLauncher addSparkArgs(String... args) {
+    for (String arg : args) {
+      sparkArgs.add(arg);
+    }
+    return this;
+  }
+
+  // Visible for testing.
+  List<String> buildSparkSubmitArgs() {
+    List<String> args = new ArrayList<String>();
+
+    if (verbose) {
+      args.add("--verbose");
+    }
+
+    if (master != null) {
+      args.add("--master");
+      args.add(master);
+    }
+
+    if (deployMode != null) {
+      args.add("--deploy-mode");
+      args.add(deployMode);
+    }
+
+    if (appName != null) {
+      args.add("--name");
+      args.add(appName);
+    }
+
+    for (Map.Entry<String, String> e : conf.entrySet()) {
+      args.add("--conf");
+      args.add(String.format("%s=%s", e.getKey(), e.getValue()));
+    }
+
+    if (propertiesFile != null) {
+      args.add("--properties-file");
+      args.add(propertiesFile);
+    }
+
+    if (!jars.isEmpty()) {
+      args.add("--jars");
+      args.add(join(",", jars));
+    }
+
+    if (!files.isEmpty()) {
+      args.add("--files");
+      args.add(join(",", files));
+    }
+
+    if (!pyFiles.isEmpty()) {
+      args.add("--py-files");
+      args.add(join(",", pyFiles));
+    }
+
+    if (userClass != null) {
+      args.add("--class");
+      args.add(userClass);
+    }
+
+    args.addAll(sparkArgs);
+    if (userResource != null) {
+      args.add(userResource);
+    }
+    args.addAll(userArgs);
+
+    return args;
+  }
+
+  @Override
+  protected List<String> buildLauncherCommand() throws IOException {
+    List<String> cmd = createJavaCommand();
+    addOptionString(cmd, System.getenv("SPARK_SUBMIT_OPTS"));
+    addOptionString(cmd, System.getenv("SPARK_JAVA_OPTS"));
+
+    // Load the properties file and check whether spark-submit will be running the app's driver
+    // or just launching a cluster app. When running the driver, the JVM's argument will be
+    // modified to cover the driver's configuration.
+    Properties props = loadPropertiesFile();
+    boolean isRunningDriver = isRunningDriver(props);
+
+    String extraClassPath = isRunningDriver ? find(DRIVER_CLASSPATH, conf, props) : null;
+    cmd.add("-cp");
+    cmd.add(join(File.pathSeparator, buildClassPath(extraClassPath)));
+
+    String libPath = null;
+    if (isRunningDriver) {
+      // Figuring out where the memory value come from is a little tricky due to precedence.
+      // Precedence is observed in the following order:
+      // - explicit configuration (setConf()), which also covers --driver-memory cli argument.
+      // - user properties, if properties file is explicitly set.
+      // - SPARK_MEM env variable
```

Comment:
doesn't `SPARK_MEM` always have lower precedence than `spark.driver.memory` regardless of how that config is set? (with or without custom properties file)